### ISSUE TITLE
allow NULL to be considered as "" if used together with a length of 0

### ISSUE THREAD

**Issue by carenas** [PhilipHazel/pcre2#54]
I usually prefer to support `ptr:NULL, length:0` for string manipulations (all `string.h` functions support it) but it looks like this is not the case for pcre2. I wouldn't mind supporting it though.

_Originally posted by @zherczeg in https://github.com/PhilipHazel/pcre2/issues/53#issuecomment-974762207_

**Comment by PhilipHazel** [comment]
The original intention for PCRE2 was that NULL always gives an error, whatever the length, for both patterns and subjects, whereas an empty string is of course a valid pattern and a valid subject. If NULL with zero length is supported as an empty subject, it should also be supported as a pattern, and as a replacement string in pcre2_substitute() (and any others I've overlooked) for consistency. Is it really worth doing this work? How much would such a feature be used?

**Comment by zherczeg** [comment]
In most system, don't supporting it requires effort, not supporting it :) Anyway, it is definitely not worth to spending much effort on it.

**Comment by ltrzesniewski** [comment]
I brought that up because in C#, `ReadOnlySpan<char>.Empty` is a struct which contains a NULL pointer and a zero length, and PCRE2 didn't like that kind of input, which surprised me. I had to [special-case](https://github.com/ltrzesniewski/pcre-net/blob/2d4c01f625e49ef0d6d9ad7f27f5d1ff236c3340/src/PCRE.NET/Internal/InternalRegex.cs#L211-L212) this in PCRE.NET, and I thought maybe other people stumbled upon it as well.

I don't know if I'd even call that a feature, it just seemed like a small inconsistency to me. I can send a PR to support this later on if you deem it's worth the trouble.

**Comment by carenas** [comment]
> The original intention for PCRE2 was that NULL always gives an error, whatever the length

I think this makes sense design wise, and the only interface that is currently missing AFAIK (other than the fixes needed so that we don't crash instead) is the use of NULL for the replacement string in pcre2_substitute, which is not documented/coded to return PCRE2_NULL_ERROR.

presume that since we will crash now instead, changing that wouldn't require changes to the API and is unlikely to get anyone surprised.

would be great if this could be tested after to avoid regressions, but I am not sure what would be the best way to do so

**Comment by PhilipHazel** [comment]
I have found (and fixed, but not yet pushed) a couple more places where tests for NULL were needed. During this process, I discovered that pcre2_jit_match(), which is "fast path with no checks use at your own risk" facility does in fact treat NULL with length zero as an empty string. With this bit of extra evidence I have now decided that I will make other functions do the same for just that special case NULL+length=0. NULL with non-zero length will still give an error - pcre2_jit_match() crashes, but that's OK because it is documented not to do sanity checks.
